File: haskell/BinarySearch.hs
binarySearchFirstTrue :: (Int -> Bool) -> Int -> Int -> Int
binarySearchFirstTrue predicate lo hi
  | lo > hi = lo
  | predicate mid = binarySearch predicate lo mid - 1
  | otherwise = binarySearch predicate mid + 1 hi
  where mid = (lo + hi) `div` 2

-- Usage example
main = print $ binarySearchFirstTrue (\x -> x >= 5) 0 10

File: python/binary_search.py
def binary_search_first_true(predicate, from_inclusive, to_inclusive):
    lo = from_inclusive - 1
    hi = to_inclusive + 1
    while hi - lo > 1:
        mid = (lo + hi) // 2
        if not predicate(mid):
            lo = mid
        else:
            hi = mid
    return hi


def test():
    assert 4 == binary_search_first_true(lambda x: x * x >= 10, 0, 10)


test()
